Failed cargo install with SSL feature without openssl3

## Summary

1. Issued command `cargo install --git https://github.com/RustPython/RustPython --features=ssl`
2. Got error while building

## System Info

* Linux: Ubuntu 20.04 LTS
* Rust: rustc 1.76.0 (07dca489a 2024-02-04)
* RustPython: 0.3.0
* Openssl: openssl (1.1.1f-1ubuntu2.22)
* `libssl-dev` (1.1.1f-1ubuntu2.22)

> Need more info? Just ask.

## Actual

```
    Updating git repository `https://github.com/RustPython/RustPython`
  Installing rustpython v0.3.0 (https://github.com/RustPython/RustPython#e6c6f966)
    Updating crates.io index
    Updating git repository `https://github.com/RustPython/Parser.git`
    Updating git repository `https://github.com/RustPython/__doc__`
  Downloaded log v0.4.21
  Downloaded syn v2.0.52
  Downloaded 2 crates (298.2 KB) in 3.34s
   Compiling version_check v0.9.4
   ( ... Shortened the long list of Downloaded and Compiling ... )
error[E0599]: no method named `all_certificates` found for reference `&X509StoreRef` in the current scope
   --> stdlib/src/ssl.rs:706:18
    |
703 |               let certs = self
    |  _________________________-
704 | |                 .ctx()
705 | |                 .cert_store()
706 | |                 .all_certificates()
    | |                 -^^^^^^^^^^^^^^^^ method not found in `&X509StoreRef`
    | |_________________|
    | 

For more information about this error, try `rustc --explain E0599`.
error: could not compile `rustpython-stdlib` (lib) due to 1 previous error
error: failed to compile `rustpython v0.3.0 (https://github.com/RustPython/RustPython#e6c6f966)`, intermediate artifacts can be found at `/tmp/cargo-installm6J4G5`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.
```

Things I've done to try to fix:

- Checked `rustup update` for rust updates (`rustc 1.76.0 (07dca489a 2024-02-04)`)
- Tried `cargo install --git https://github.com/RustPython/RustPython` (non-SSL works fine)
- Tried [d0c827ed] with SSL (no luck)

> If anyone could offer further suggestions.

`cargo vcpkg build` seems to install openssl:

```
    Fetching vcpkg
    Checkout rev 2024.02.14
  Installing openssl
    Finished in 37.78s
```

but `cargo build --release --features ssl` fails with the same error:

```
error[E0599]: no method named `all_certificates` found for reference `&X509StoreRef` in the current scope
   --> stdlib\src\ssl.rs:706:18
    |
703 |               let certs = self
    |  _________________________-
704 | |                 .ctx()
705 | |                 .cert_store()
706 | |                 .all_certificates()
    | |                 -^^^^^^^^^^^^^^^^ method not found in `&X509StoreRef`
    | |_________________|
    |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `rustpython-stdlib` (lib) due to 1 previous error
```

--- a/stdlib/build.rs
+++ b/stdlib/build.rs
@@ -1,29 +1,34 @@
 fn main() {
     println!(r#"cargo::rustc-check-cfg=cfg(osslconf, values("OPENSSL_NO_COMP"))"#);
-    println!("cargo::rustc-check-cfg=cfg(ossl101)");
-    println!("cargo::rustc-check-cfg=cfg(ossl102)");
-    println!("cargo::rustc-check-cfg=cfg(ossl110)");
-    println!("cargo::rustc-check-cfg=cfg(ossl110g)");
-    println!("cargo::rustc-check-cfg=cfg(ossl111)");
+
+    #[allow(clippy::unusual_byte_groupings)]
+    let ossl_vers = [
+        (0x1_00_01_00_0, "ossl101"),
+        (0x1_00_02_00_0, "ossl102"),
+        (0x1_01_00_00_0, "ossl110"),
+        (0x1_01_00_07_0, "ossl110g"),
+        (0x1_01_00_08_0, "ossl110h"),
+        (0x1_01_01_00_0, "ossl111"),
+        (0x1_01_01_04_0, "ossl111d"),
+        (0x3_00_00_00_0, "ossl300"),
+        (0x3_01_00_00_0, "ossl310"),
+        (0x3_02_00_00_0, "ossl320"),
+        (0x3_03_00_00_0, "ossl330"),
+    ];
+
+    for (_, cfg) in ossl_vers {
+        println!("cargo::rustc-check-cfg=cfg({cfg})");
+    }
+
     #[allow(clippy::unusual_byte_groupings)]
     if let Ok(v) = std::env::var("DEP_OPENSSL_VERSION_NUMBER") {
         println!("cargo:rustc-env=OPENSSL_API_VERSION={v}");
         // cfg setup from openssl crate's build script
         let version = u64::from_str_radix(&v, 16).unwrap();
-        if version >= 0x1_00_01_00_0 {
-            println!("cargo:rustc-cfg=ossl101");
-        }
-        if version >= 0x1_00_02_00_0 {
-            println!("cargo:rustc-cfg=ossl102");
-        }
-        if version >= 0x1_01_00_00_0 {
-            println!("cargo:rustc-cfg=ossl110");
-        }
-        if version >= 0x1_01_00_07_0 {
-            println!("cargo:rustc-cfg=ossl110g");
-        }
-        if version >= 0x1_01_01_00_0 {
-            println!("cargo:rustc-cfg=ossl111");
+        for (ver, cfg) in ossl_vers {
+            if version >= ver {
+                println!("cargo:rustc-cfg={cfg}");
+            }
         }
     }
     if let Ok(v) = std::env::var("DEP_OPENSSL_CONF") {

--- a/stdlib/src/ssl.rs
+++ b/stdlib/src/ssl.rs
@@ -700,12 +700,14 @@ mod _ssl {
             vm: &VirtualMachine,
         ) -> PyResult<Vec<PyObjectRef>> {
             let binary_form = binary_form.unwrap_or(false);
-            let certs = self
-                .ctx()
-                .cert_store()
-                .all_certificates()
-                .iter()
-                .map(|cert| cert_to_py(vm, cert, binary_form))
+            let ctx = self.ctx();
+            #[cfg(ossl300)]
+            let certs = ctx.cert_store().all_certificates();
+            #[cfg(not(ossl300))]
+            let certs = ctx.cert_store().objects().iter().filter_map(|x| x.x509());
+            let certs = certs
+                .into_iter()
+                .map(|ref cert| cert_to_py(vm, cert, binary_form))
                 .collect::<Result<Vec<_>, _>>()?;
             Ok(certs)
         }

